### PR TITLE
Added logging of errors during threadpool operations.

### PR DESCRIPTION
**Pull request by larsgeorge-db**
Fixes: #89

### CODE REVIEW

File: src/databricks/labs/ucx/framework/parallel.py [hunk 1]
```diff
@@ -45,22 +48,36 @@ def __init__(
...
         logger.debug(f"Starting {len(self._executables)} tasks in {self._num_threads} threads")
 
+        def error_handler(func, name):
+            @functools.wraps(func)
+            def inner(*args, **kwargs):
+                try:
+                    return func(*args, **kwargs)
+                except Exception as ex:
+                    msg = getattr(ex, "message", ex)
+                    logger.warning(f"An error occurred while '{name}', message: {msg}")
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/parallel.py:63]:
```suggestion
                    logger.warning(f"{name} task failed: {str(ex)}")
```
python exceptions don't always have `message`. and we need more concise log messages.

File: src/databricks/labs/ucx/framework/parallel.py [hunk 2]
```diff
@@ -45,22 +48,36 @@ def __init__(
...
+                future = executor.submit(error_handler(executable, self._name))
                 future.add_done_callback(self._done_callback)
-                # TODO: errors are not handled yet - https://github.com/databricks/UC-Upgrade/issues/89
                 self._futures.append(future)
 
             results = concurrent.futures.wait(self._futures, return_when=ALL_COMPLETED)
 
         logger.debug("Collecting the results from threaded execution")
-        collected = [future.result() for future in results.done]
+        collected = [future.result() for future in results.done if future.result()]
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/parallel.py:77]: i think there's a bug here, as `.result()` may not always be available the second time we call it. Just for counting, introduce a separate variable.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/framework/parallel.py (hunk 2 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/framework/parallel.py:77]: Really? I thought (by reading the `_base.py` code) that only reading exceptions has side effects?

File: src/databricks/labs/ucx/framework/parallel.py (hunk 1 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/framework/parallel.py:63]: But that is why I have the `getattr()` with the default value. Why call `str()` on the exception, a simple `{ex}` does the same, no (that is, call `__str__`)?

**Comment by larsgeorge-db** [comment]
```
============================= test session starts ==============================
collecting ... collected 1 item
21:35  WARN [d.l.ucx.framework.parallel] testing task failed: failed
21:35  WARN [d.l.ucx.framework.parallel] The parallel run of 'testing' incurred 1 failed tasks

unit/framework/test_parallel.py::test_threaded_execution_gather_with_failed_task PASSED

============================== 1 passed in 0.22s ===============================

Process finished with exit code 0
```

**Review by nfx — APPROVED** [review]

**Pull request merged**